MAINT: Simplify logic in convert_datetime_to_datetimestruct

eric-wieser opened the pull request:
Most of this is just emulating the behavior of python's divmod. With a carefully crafted out parameter, we can eliminate a lot of repetition.

eric-wieser committed:
MAINT: Simplify logic in convert_datetime_to_datetimestruct

Most of this is just emulating the behavior of python's divmod. With a carefully crafted out parameter, we can eliminate a lot of repetition

eric-wieser commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -473,169 +486,100 @@ convert_datetime_to_datetimestruct(PyArray_DatetimeMetaData *meta,
+            set_datetimestruct_days(extract_unit(&dt, perday), out);
+            out->hour = (int)extract_unit(&dt, 60*60);
+            out->min  = (int)extract_unit(&dt, 60);
+            out->sec  = (int)dt;
```
Looks even tidier as:
```
out->sec  = (int)extract_unit(&dt, 60);
out->min  = (int)extract_unit(&dt, 60);
out->hour = (int)extract_unit(&dt, 24);
set_datetimestruct_days(dt, out);
```
But that's less efficient, as the `dt < 0` branch can't be optimized out in the non-first calls to `extract_unit`

tylerjereddy replied:
I'm not too worried about efficiency just yet myself--we both know there's a lot more work to do still for correctness.

My only reservation with refactoring / abstracting is that we can't trust our current tests to catch subtle (and even breaking) changes just yet.

eric-wieser committed:
TST: add tests for the existence and round-tripping of datetime limits

tylerjereddy reviewed (commented):
In #13168 you note:

> I'll make a separate PR to apply the above cleanup, without doing any attempts at preventing overflow

But here you've done better than that -- you've replaced the atrocious `dt - (perday-1)` for `dt < 0` cases with an early integer division of the time delta in `extract_unit()`.

In fact, I checked out your branch from here, selectively pulled in my overflow unit test from the cognate PR, removed the "weeks" units, and it passed for all other cases. Indeed, I suggest that may be worth doing, or I can push it in if you want.

That way maybe we're both happy--I don't muck up the integer arithmetic, and there's a test that scans over most of the units to start enforcing sensible properties for our datetime spans, which fails more broadly on master.

eric-wieser commented:
> I can push it in if you want.

Feel free

eric-wieser commented:
I applied my own comments to your test, hope that's ok.

Interestingly the test passes just fine on weeks for me.

tylerjereddy commented on numpy/core/tests/test_datetime.py:
```diff
@@ -2220,6 +2221,38 @@ def test_corecursive_input(self):
+
+        # finish the roundtrip to unix epoch:
+        restored = inv_op(limit, max_delta)
+        assert restored == epoch
```
This isn't as sensitive as what I had before. I can restore the sensitivity to issues with weeks:

```diff
--- a/numpy/core/tests/test_datetime.py
+++ b/numpy/core/tests/test_datetime.py
@@ -2237,20 +2237,15 @@ class TestDateTime(object):
         """
         # TODO: add absolute (gold standard) time span limits
         epoch = np.datetime64(0, time_unit)
-        max_delta = np.timedelta64(np.iinfo(np.int64).max, time_unit)
-
-        # subtraction is particularly prone to overflow in older code, making
-        # this fail
-        limit = op(epoch, max_delta)
-        assert cmp_op(limit, epoch)
+        max_delta = np.timedelta64(np.iinfo(np.int64).max)
 
         # The overflow would show up in the repr so test a roundtrip from epoch
         # to limit and back using the str representation
-        limit_via_str = np.datetime64(str(limit))
-        assert limit_via_str == limit
+        limit_via_str = np.datetime64(str(op(epoch, max_delta)))
+        assert cmp_op(limit_via_str, epoch)
 
         # finish the roundtrip to unix epoch:
-        restored = inv_op(limit, max_delta)
+        restored = inv_op(limit_via_str, max_delta)
         assert restored == epoch

```

I can understand some desire for being explicit with the `m8` units, but it should be perfectly valid to add generic deltas to datetime64 values in the respective units from each parametrization. Being implicit increases sensitivity if you also use the string path on the final assert.

eric-wieser replied:
Something isn't making sense here. I'll have a play around with the union of our two tests

tylerjereddy replied:
```python
ref = np.datetime64(0, 'W')
max_delta = np.timedelta64(np.iinfo(np.int64).max)
max_delta_units = np.timedelta64(np.iinfo(np.int64).max, 'W')
generic_round_trip = np.datetime64(np.datetime64(str(ref - max_delta)) + max_delta) 
units_round_trip = np.datetime64(np.datetime64(str(ref - max_delta_units)) + max_delta_units)
print(generic_round_trip) # 1970-01-07
print(units_round_trip) #1970-01-01
```

eric-wieser replied:
Thanks for the easily-pasteable example

That test isn't valid, and produces different results without overflow:
```
ref = np.datetime64(0, 'W')
delta = np.timedelta64(1)
delta_unit = np.timedelta64(1, 'W')
ref_plus_delta = ref + delta
assert ref_plus_delta == ref + delta_unit  # makes no difference here

# this converts from `M8[W]` to `M8[D]
ref_plus_delta_str = np.datetime64(str(ref_plus_delta))

r1 = ref_plus_delta_str - delta_unit  # subtracts weeks
r2 = ref_plus_delta_str - delta  # subtracts days, since the unit changed
```

What you've proven there is that:
* generic timedeltas are fragile
* Round-tripping via string might change the unit

tylerjereddy replied:
I still think the test was just fine as it was--unless those fragilities / behaviors are intentional it seems like enforcing the expectation that those tests pass is reasonable, with a skip on the weeks for now.

If it is just weeks that magically switches like that it seems like something to be addressed for weeks and locked down for the other units in the test.

tylerjereddy replied:
Maybe even add another assertion on the `m8` unit of time preservation through the string conversion.

eric-wieser replied:
> Maybe even add another assertion on the m8 unit of time preservation through the string conversion.

Sounds like a good idea for a separate test.

> unless those fragilities / behaviors are intentional

They are. There's no way to tell from `1970-01-01` if the user wanted `W`, `D`, `2D`, etc - we choose a unit based on the precision of the date string, and we have no prevision between `YYYY-MM` and `YYYY-MM-DD`

tylerjereddy replied:
If you convert from weeks to days near the integral size limit you have an impossible decision to make, right? What year is it? We can't balance the magnitude of the date with the precision of the unit. I can't think of a sensible reason we should allow that or how it would behave logically.

We could probably agree that the unit test for round tripping symmetry is just tying to do too much if you throw in the generic `m8` issues as well. It may be rather verbose to add a second parametrized test over all the units for that, but fine by me if you prefer that.

eric-wieser replied:
> I can't think of a sensible reason we should allow that or how it would behave logically.

In my opinion, `np.datetime64("ONLY_FITS_IN_WEEKS-01-01")` should error / give NaT, not retry with weeks if days do not fit.

Remember the user can always ask for  `np.datetime64("ONLY_FITS_IN_WEEKS-01-01", 'W')`

eric-wieser commented:
Tests updated.

Roundtripping via `np.datetime64(str(dt), np.datetime_unit(dt))` was sufficient to test those codepaths.

I _was_ able to replicate a `unit='W'` failure though, so I've added that to the suite as an `xfail`. It fails on units like `10D` too, which is not surprising.

I removed the dance of addition and subtraction - it's not interesting, because it just exercises integer addition under the hood.

tylerjereddy reviewed (approved):
Ok, I confirmed that `test_limit_str_roundtrip` detects multiple failures when cherry-picked into master branch, all of them in the negative time span direction, which was the original thing I was trying to improve.

Ultimately, having current tests pass + new properties enforced with new tests is what I was looking for for datetime here.

tylerjereddy commented:
Thanks @eric-wieser  

Looks like I lost credit for drafting the tests, but there are bigger problems in the world :)

eric-wieser commented:
You'll have to work out how to fix the weeks so you can take credit for that instead :).

Solving the 10D problem is probably not worth the effort, and seems unsolvable in any sensible way.

For things like `100Y` as a unit, the result is not even representable.

Perhaps the thing to do is just detect the overflow and error.
